extmod/vfs_posix: Fix relative paths on non-root VFS

cwalther opened the pull request:
`VfsPosix` has a couple of bugs regarding handling of relative paths and `chdir()`/`getcwd()` that surface when it is instanced with a non-empty root path (i.e. only a subtree of the outer filesystem is exposed to Python).

They are somewhat entangled, but I tried my best to separate the fixes into individually reviewable commits, each with first a failing test that demonstrates what is going wrong, followed by fixes that makes it pass. (The tricky part of that was to get the fixes in the right order and craft the tests such that they actually pass after the respective fix, rather than continuing to fail due to the remaining bugs.)

1. A `VfsPosix` created with a relative root path gets confused when `chdir()` is called on it and becomes unable to properly resolve absolute paths, because changing directories effectively shifts its root. The simplest fix for that would be to say “don’t do that”, but since the unit tests themselves do it, fix it by making a relative path absolute before storing it.

2. The unwritten API contract expected of a VFS by `mp_vfs_lookup_path()` is that paths passed in are relative to the root of the VFS if they start with `/` and relative to the current directory of the VFS otherwise. This is not correctly implemented in `VfsPosix` for instances with a non-empty root – all paths are interpreted relative to the root. Fix that. Since `VfsPosix` tracks its CWD using the “external” CWD of the Unix process, the correct handling for relative paths is to pass them through unmodified.

3. The unwritten API contract expected of a VFS`.getcwd()` by `mp_vfs_getcwd()` is that its return value should be either `""` or `"/"` when the CWD is at the root of the VFS and otherwise start with a slash and not end with a slash. This is not correctly implemented in `VfsPosix` for instances with a non-empty root – the required leading slash, if any, is cut off because the root length includes a trailing slash. This results in missing slashes in the middle of the return value of `os.getcwd()` or in uninitialized garbage from beyond a string’s null terminator when the CWD is at the VFS root.

The matter is complicated by the fact that there are some faulty existing tests that only pass by accident due to bug 2. Unmodified, they would fail after the fix for bug 2 (5th commit), however that is insignificant because they actually test an unrealistic situation. Fixing them in the 3rd commit to match the realistic situation makes them pass with bug 2 fixed (and also without, when done in this order, because after the 2nd commit, bug 2 has no effect in that case).

The unrealistic situation is that `VfsPosix` methods are called with relative paths while the current working directory is somewhere outside of the root of the VFS. In the intended use of VFS objects via `os.mount()` (as opposed to calling methods directly as the tests do), this never happens, as `mp_vfs_lookup_path()` directs incoming calls to the VFS that contains the CWD. Fix this by explicitly changing the working directory to the root of the VFS before calling methods on it, as the subsequent relative path accesses expect.

cwalther committed:
extmod/vfs_posix: Fix relative paths on non-root VFS.

The unwritten API contract expected of a VFS by mp_vfs_lookup_path() is
that paths passed in are relative to the root of the VFS if they start
with '/' and relative to the current directory of the VFS otherwise.
This was not correctly implemented in VfsPosix for instances with a
non-empty root - all paths were interpreted relative to the root. Fix
that. Since VfsPosix tracks its CWD using the "external" CWD of the Unix
process, the correct handling for relative paths is to pass them through
unmodified.

Also, when concatenating absolute paths, fix an off-by-one resulting in
a harmless double slash (the root path already has a trailing slash).

Signed-off-by: Christian Walther <cwalther@gmx.ch>

cwalther committed:
extmod/vfs_posix: Fix getcwd() on non-root VFS.

The unwritten API contract expected of a VFS.getcwd() by mp_vfs_getcwd()
is that its return value should be either "" or "/" when the CWD is at
the root of the VFS and otherwise start with a slash and not end with a
slash. This was not correctly implemented in VfsPosix for instances with
a non-empty root - the required leading slash, if any, was cut off
because the root length includes a trailing slash. This would result in
missing slashes in the middle of the return value of os.getcwd() or in
uninitialized garbage from beyond a string's null terminator when the
CWD was at the VFS root.

Signed-off-by: Christian Walther <cwalther@gmx.ch>

cwalther committed:
extmod/vfs_posix: Fix accidentally passing tests.

These tests test an unrealistic situation and only pass by accident due
to a bug. The upcoming fix for the bug would make them fail.

The unrealistic situation is that VfsPosix methods are called with
relative paths while the current working directory is somewhere outside
of the root of the VFS. In the intended use of VFS objects via
os.mount() (as opposed to calling methods directly as the tests do),
this never happens, as mp_vfs_lookup_path() directs incoming calls to
the VFS that contains the CWD.

Make the testing situation realistic by changing the working directory
to the root of the VFS before calling methods on it, as the subsequent
relative path accesses expect.

Thanks to the preceding commit, the tests still pass, but still for the
wrong reason. The following commit "Fix relative paths on non-root VFS"
will make them pass for the correct reason.

Signed-off-by: Christian Walther <cwalther@gmx.ch>

cwalther committed:
extmod/vfs_posix: Fix relative root path.

A VfsPosix created with a relative root path would get confused when
chdir() was called on it and become unable to properly resolve absolute
paths, because changing directories effectively shifted its root. The
simplest fix for that would be to say "don't do that", but since the
unit tests themselves do it, fix it by making a relative path absolute
before storing it.

Signed-off-by: Christian Walther <cwalther@gmx.ch>

cwalther commented:
I’ll take a look at the Windows failures tomorrow (I didn’t expect having to care about Windows in something named “posix”). The coverage failure seems out of my hands.

dpgeorge commented on extmod/vfs_posix.c:
```diff
@@ -107,7 +112,28 @@ STATIC mp_obj_t vfs_posix_make_new(const mp_obj_type_t *type, size_t n_args, siz
+            char buf[MICROPY_ALLOC_PATH_MAX + 1];
+            const char *cwd = getcwd(buf, sizeof(buf));
+            if (cwd == NULL) {
+                mp_raise_OSError(errno);
```
Are you able to write a test which exercises this line?

dpgeorge replied:
I think the only options are a path that is too long, or a path that the program doesn't have permission to access.  For the latter, maybe `micropython` can be executed when in a root-only directory like `/root`?

cwalther replied:
A long path is the only case I can think of that works without too many platform-specific hoops to jump through. In the additional commit 6815f70, I got it working locally on macOS and Linux. Not on Windows, so the test is skipped there. I’m not sure what’s going on with the CI failures, these seem totally random. Any ideas, before I spend a lot of effort trying to debug something I can’t reproduce locally?

cwalther replied:
Actually, now that I think of it, now that Windows is out anyway (where this wouldn’t work), I could use the failure mode “[ENOENT] A component of the pathname no longer exists.” instead of the too long path. Tried that in e78dc9a, let’s see if that works better. (Before merging, the last two commits should be squashed – I only did it as an additional commit rather than a force-push because I think otherwise the mysterious CI results of 6815f70, which you may still want to look at, might be lost.)

dpgeorge replied:
That's great you could write such a test, very good thank you!

The CI failure you saw is because `raise` with no arguments is not supported by the native emitter.  Instead do `raise er`.

cwalther replied:
Thanks, that did indeed fix not only the `emit_native_raise_varargs: Assertion 'n_args == 1' failed.` that sounds clear enough in hindsight, but also the more mysterious `TypeError: unsupported types for __lt__: 'dict', 'int'` when compiled without assertions. I had missed that those failures occurred with `--emit native` and could reproduce them that way.

So with that fix, both tests would work. Let me know which one you prefer. The newer one with the deleted directory is probably faster and less susceptible to unexpected external circumstances than the one that creates directories with extremely long names.

dpgeorge replied:
Yes, the newer one is much better, thanks!

cwalther committed:
extmod/vfs_posix: Additional tests for coverage of error cases.

Signed-off-by: Christian Walther <cwalther@gmx.ch>

dpgeorge commented on tests/extmod/vfs_posix_nonwindows.py:
```diff
@@ -0,0 +1,32 @@
+# Test for VfsPosix error conditions that cannot be reproduced on Windows
```
I think this test should be called `vfs_posix_enoent.py`.  It may eventually run on windows, or may not run on other systems eg macOS.

cwalther replied:
Done. (It does currently run on macOS, that’s where I’m developing.)

dpgeorge commented on tests/extmod/vfs_posix_nonwindows.py:
```diff
@@ -0,0 +1,32 @@
+    raise SystemExit
+
+curdir = os.getcwd()
+os.mkdir("aaa")
```
I suggest to use a more unique name, eg "vfs_posix_test_dir".

cwalther replied:
Done. I chose “vfs_posix_enoent_test_dir” in analogy with the vfs_posix_ilistdir_* tests, and also added the skip-if-already-exists from there.

cwalther commented:
Addressed review comments and rebased on master (which required no changes).

dpgeorge commented on tests/extmod/vfs_posix.py:
```diff
@@ -130,6 +130,70 @@ def write_files_without_closing():
 except OSError:
     print("remove OSError")
 
+# construct new VfsPosix with absolute path argument
```
This commit "Add more tests for relative paths, currently failing." adds a lot to the existing test file and makes it too complex.  So please can you put these new tests in a separate test file, maybe `vfs_posix_paths.py`.

Also, we usually have tests in the same commit as the commit that fixes the issue.  So please also try to do that here if it makes sense (I understand you made changes to tests that were accidentally passing, and those can indeed stay in a separate commit.).

cwalther replied:
OK, will do. Probably not before the weekend.

(In case you were wondering about the separate commits, there is a reason I did that. I always do it that way, because it makes it easier to check that 1. something is wrong (by checking out the first commit with the failing test) and 2. the change actually fixes it (by checking out the second commit with the fix). But of course I willl conform to the project style.)

dpgeorge replied:
I'm fully on board with making lots of small commits and separating out functionality into different commits.  In general we try to do that in this repo.

But for this particular case, I felt that it was a bit confusing which test went with which fix.  And I think it would be easier for someone reading the git history if the fix and test were in the same commit.

cwalther replied:
Done. Apologies it took me several weekends.

dpgeorge replied:
No problem at all.

cwalther commented:
- Split added tests out into a new file `vfs_posix_paths.py`.
- Squashed test and corresponding fix into the same commit.
- Rebased on master (which required no changes).

dpgeorge commented:
Thanks for updating and making clean and independent commits.  Merged as-is.

cwalther commented:
Thank you!
